File: src/main/java/com/spagnola/lcars/elements/LCARSTime.java
/**
 *
 */
package com.spagnola.lcars.elements;


import com.spagnola.lcars.LCARS;

import java.awt.FontMetrics;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.TimeZone;
import java.util.Timer;
import java.util.TimerTask;

/**
 * @author Perry Spagnola
 *
 */
public class LCARSTime extends LCARSLabel {

    public static final int DEFAULT = 0;
    public static final int DAY_OF_WEEK = 1;
    public static final int AM_PM = 2;
    public static final int DATE = 3;
    public static final int TIME_ZONE = 4;
    public static final int TIME_ZONE_RFC822 = 5;
    public static final int TIME = 6;
    public static final int SECONDS = 7;
    public static final int TIME_24 = 8;
    protected static final SimpleDateFormat[] formattedTime = {
            new SimpleDateFormat("HH:mm:ss"),
            new SimpleDateFormat("EEEE"),
            new SimpleDateFormat("a"),
            new SimpleDateFormat("MMMM d, yyyy"),
            new SimpleDateFormat("zzzz"),
            new SimpleDateFormat("Z"),
            new SimpleDateFormat("h:mm"),
            new SimpleDateFormat("ss"),
            new SimpleDateFormat("HH:mm")
    };
    protected static TimeZone timezone;
    protected SimpleDateFormat format;

    /**
     * @param parent
     * @param x
     * @param y
     * @param style
     */
    public LCARSTime(int x, int y, int style) {
        super("", x, y, style);

        format = formattedTime[DEFAULT];

        init();

    }

    protected void init() {
        componentText = format.format(Calendar.getInstance().getTime());
        timezone = Calendar.getInstance().getTimeZone();

        FontMetrics fm = getFontMetrics(LCARSFont.deriveFont(fontSize));
        w = fm.stringWidth(componentText);
        h = fm.getHeight();
        createLabelShape();

        setBounds(0, 0, w, h);

        setTextInsets(0, -10);
        //setForeground(LCARS.getLCARSColor(style));

        (new Timer()).schedule(new TimerTask() {
            @Override
            public void run() {
                format.setTimeZone(timezone);
                setText(format.format(Calendar.getInstance().getTime()));
                repaint();
            }
        }, LCARS.SECOND, LCARS.SECOND);
    }


    public LCARSTime(int x, int y, int mode, int size, int style) {
        super("", x, y, style);

        fontSize = size;

        format = formattedTime[mode];

        init();

    }

    public static TimeZone getTimeZone() {
        return timezone;
    }

    public static void setTimeZone(String timezoneID) {
        timezone = TimeZone.getTimeZone(timezoneID);
    }

    public static void setTimeZone() {
        timezone = Calendar.getInstance().getTimeZone();
    }

    public void setMode(int mode) {
        format = formattedTime[mode];
    }
}

File: src/main/kotlin/de/visualdigits/lcars/elements/LCARSTime.kt
/**
 *
 */
package de.visualdigits.lcars.elements


import de.visualdigits.lcars.LCARS
import java.awt.FontMetrics
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.TimeZone
import java.util.Timer
import java.util.TimerTask

/**
 * @author Perry Spagnola
 */
class LCARSTime : de.visualdigits.lcars.elements.LCARSLabel {

    private var format: SimpleDateFormat

    /**
     * @param parent
     * @param x
     * @param y
     * @param style
     */
    constructor(x: Int, y: Int, style: Int) : super("00:00", x, y, style) {
        format = de.visualdigits.lcars.elements.LCARSTime.Companion.formattedTime[de.visualdigits.lcars.elements.LCARSTime.Companion.DEFAULT]

        init()
    }

    private fun init() {
        componentText = format.format(Calendar.getInstance().time)
        de.visualdigits.lcars.elements.LCARSTime.Companion.timeZone = Calendar.getInstance().getTimeZone()

        val fm: FontMetrics = getFontMetrics(LCARSFont.deriveFont(fontSize))
        w = fm.stringWidth(componentText)
        h = fm.height
        createLabelShape()

        setBounds(0, 0, w, h)

        setTextInsets(0, -10)

        //setForeground(LCARS.getLCARSColor(style));
        (Timer()).schedule(object : TimerTask() {
            override fun run() {
                format.timeZone = de.visualdigits.lcars.elements.LCARSTime.Companion.timeZone
                componentText = format.format(Calendar.getInstance().time)
                repaint()
            }
        }, LCARS.SECOND.toLong(), LCARS.SECOND.toLong())
    }


    constructor(x: Int, y: Int, mode: Int, size: Int, style: Int) : super("00:00", x, y, style) {
        fontSize = size.toFloat()

        format = de.visualdigits.lcars.elements.LCARSTime.Companion.formattedTime[mode]

        init()
    }

    fun setMode(mode: Int) {
        format = de.visualdigits.lcars.elements.LCARSTime.Companion.formattedTime[mode]
    }

    companion object {
        const val DEFAULT: Int = 0
        const val DAY_OF_WEEK: Int = 1
        const val AM_PM: Int = 2
        const val DATE: Int = 3
        const val TIME_ZONE: Int = 4
        const val TIME_ZONE_RFC822: Int = 5
        const val TIME: Int = 6
        const val SECONDS: Int = 7
        const val TIME_24: Int = 8
        private val formattedTime: Array<SimpleDateFormat> = arrayOf<SimpleDateFormat>(
            SimpleDateFormat("HH:mm:ss"),
            SimpleDateFormat("EEEE"),
            SimpleDateFormat("a"),
            SimpleDateFormat("MMMM d, yyyy"),
            SimpleDateFormat("zzzz"),
            SimpleDateFormat("Z"),
            SimpleDateFormat("h:mm"),
            SimpleDateFormat("ss"),
            SimpleDateFormat("HH:mm")
        )
        var timeZone: TimeZone? = Calendar.getInstance().getTimeZone()

        fun setTimeZone(timezone: String? = null) {
            de.visualdigits.lcars.elements.LCARSTime.Companion.timeZone = timezone?.let { TimeZone.getTimeZone(it) } ?: Calendar.getInstance().getTimeZone()
        }
    }
}
